### PR TITLE
improvements to A_mul_B! section of the manual

### LINKED ISSUE #16425: Export At_ldiv_B!
Should `At_ldiv_B!` be exported from Base?  `A_ldiv_B!` and friends are, but the transposed solve method is missing.

### PR DESCRIPTION
**Pull request by stevengj**
A student of mine mentioned to me that he had trouble finding the documentation for `A_mul_B!` in the manual because it was listed under "math" and not "linear algebra".  In this PR, I fixed that and then made a few related changes:
- `A_ldiv_B!` was exported but not documented, so I added documentation and also added an optional 3-argument version to make it act a bit more like `A_mul_B!`
- `At_ldiv_B` was exported, but not `At_ldiv_B!`, as noted in JuliaParallel/PETSc.jl#70 (closes #16425).  Similarly for `Ac_ldiv_B!`.   I understand that @andreasnoack has some longer-term plans for `ldiv!`, but in the meantime the current state of affairs seems pretty inconsistent.
- Other minor documentation improvements (e.g. the BLAS section of the manual no longer assumes that the reader knows what BLAS is).

### CODE REVIEW

File: doc/stdlib/linalg.rst [hunk 1]
```diff
@@ -1311,17 +1311,180 @@ Linear algebra functions in Julia are largely implemented by calling functions f
...
 
 .. module:: Base.LinAlg.BLAS
 
-:mod:`Base.LinAlg.BLAS` provides wrappers for some of the BLAS functions for
-linear algebra.  Those BLAS functions that overwrite one of the input
-arrays have names ending in ``'!'``.
-
-Usually a function has 4 methods defined, one each for ``Float64``,
-``Float32``, ``Complex128`` and ``Complex64`` arrays.
+In Julia and most other scientific computing, linear-algebra operations are
```
Comment (tkelman) [inline doc/stdlib/linalg.rst]: most other scientific computing ...

environments? languages?

Comment (stevengj) [inline doc/stdlib/linalg.rst]: Actually, I meant "scientific computing" in general, though I should qualify that as dense linear algebra only.

Comment (tkelman) [inline doc/stdlib/linalg.rst]: Unless you happen to use eigen?

Comment (stevengj) [inline doc/stdlib/linalg.rst]: "Most," not all.   Sure, there are Eigen, Blaze, and a few others, but these are by far in the minority as far as I can tell.

Comment (stevengj) [inline doc/stdlib/linalg.rst]: I changed it to "In Julia (as in much of scientific computation), dense linear-algebra ..."

**Comment by stevengj** [comment]
AppVeyor failure seems unrelated; some kind of timeout.

**Comment by stevengj** [comment]
@andreasnoack, do you have an opinion on this?  I'm not sure what your `ldiv!` plans are or when they are likely to happen.

**Comment by andreasnoack** [comment]
All of this is great but it also conflicts (in the git sense) with https://github.com/JuliaLang/julia/pull/6837 which I plan to finish in this cycle so, from my point of view, I'd prefer to have some kind of this added after #6837 is merged. It will probably also be simpler when there is only `ldiv!`.

**Comment by andreasnoack** [comment]
After other array changes, I'll have to do a big rebase anyway so if you can rebase this I think we should merge soon.

**Comment by stevengj** [comment]
Rebased.

File: doc/stdlib/linalg.rst [hunk 2]
```diff
@@ -1459,17 +1459,180 @@ Linear algebra functions in Julia are largely implemented by calling functions f
...
+allow you to supply a pre-allocated output vector or matrix.  This is useful
+when optimizing critical code in order to avoid the overhead of repeated allocations.
+These in-place operations are suffixed with ``!`` below (e.g. ``A_mul_B!``)
+according to the usual Julia convention.
+
+.. function:: A_ldiv_B!([Y,] A, B) -> Y
+
+   .. Docstring generated from Julia source
+
+   Compute ``A  B`` in-place and store the result in ``Y``\ , returning the result. If only two arguments are passed, then ``A_ldiv_B!(A, B)`` overwrites ``B`` with the result.
```
Comment (tkelman) [inline doc/stdlib/linalg.rst]: genstdlib ate the backslash here

File: base/linalg/factorization.jl [hunk 3]
```diff
@@ -41,3 +41,41 @@ for (f1, f2) in ((:\, :A_ldiv_B!),
...
+    A_ldiv_B!([Y,] A, B) -> Y
+
+Compute `A \ B` in-place and store the result in `Y`, returning the result.
+If only two arguments are passed, then `A_ldiv_B!(A, B)` overwrites `B` with
+the result.
+
+The argument `A` should *not* be a matrix.  Rather, instead of matrices it should be a
+factorization object (e.g. produced by [`factorize`](:func:`factorize`) or [`cholfact`](:func:`cholfact`)).
+The reason for this is that factorization itself is both expensive and typically allocates memory
+(although it can also be done in-place via, e.g., [`lufact`](:func:`lufact`)),
```
Comment (tkelman) [inline base/linalg/factorization.jl]: `lufact!` ?

**Pull request merged**